### PR TITLE
Feature/extract config out of constructir

### PR DESCRIPTION
**Pull request by Leooo**
@samselikoff, first try for https://github.com/samselikoff/ember-cli-mirage/issues/842 is this what you want? Started adding a few tests.

It contains the changes in https://github.com/samselikoff/ember-cli-mirage/pull/846 too, I expect this PR to be merged soon right?

L

### CODE REVIEW

**Comment by Leooo** [comment]
@samselikoff not sure with the travis build failed on this one, tests are all passing on my side.\
L

File: addon/server.js [hunk 1]
```diff
@@ -106,45 +107,59 @@ function extractRouteArguments(args) {
...
 
     this._defineRouteHandlerHelpers();
 
-    this.db = new Db();
-    this.schema = new Schema(this.db, options.models);
-    this.serializerOrRegistry = new SerializerRegistry(this.schema, options.serializers);
+    this.db = this.db || new Db();
+    if (this.schema) {
+      this.schema.registerModels(config.models);
+      this.serializerOrRegistry.schema = this.schema;
```
Comment (samselikoff) [inline addon/server.js]: shouldn't need this line, `serializerOrRegistry` is a singleton that should already have a reference to `schema`

File: addon/server.js [hunk 2]
```diff
@@ -106,45 +107,59 @@ function extractRouteArguments(args) {
...
+      this.serializerOrRegistry._serializerMap = config.serializers || {};
```
Comment (samselikoff) [inline addon/server.js]: instead, let's add an idempotent `registerSerializers` method to the `SerializerRegistry` class that takes in a hash of serializers and augments its registry

File: addon/server.js [hunk 3]
```diff
@@ -196,7 +211,8 @@ export default class Server {
   */
   loadFactories(factoryMap) {
     // Store a reference to the factories
-    this._factoryMap = factoryMap;
+    this._factoryMap = this._factoryMap || {};
+    this._factoryMap = Object.assign(this._factoryMap, factoryMap);
```
Comment (samselikoff) [inline addon/server.js]: perfect. Could we communicate this a bit more clearly, perhaps

```js
let currentFactoryMap = this._factoryMap || {};
this._factoryMap = Object.assign(currentFactoryMap, factoryMap);
```

Comment (samselikoff) [inline addon/server.js]: Just saw the failure. If you click `details` next to the red Travis failure you can see the build logs. Search `"not ok"` and you'll see the failure, `Object.assign is not a function`. This is because we don't have a polyfill and it's not supported in Phantomjs.

[Elsewhere](https://github.com/samselikoff/ember-cli-mirage/blob/2c31ad15a46e90b357a83b6896c6774fa42c6488/addon/serializer.js#L3) we just use _.assign from lodash, so let's do that here.

```js
import _assign from 'lodash/object/assign';
...
_assign(currentFactoryMap, factoryMap)
```

**Comment by Leooo** [comment]
@samselikoff done

**Comment by Leooo** [comment]
@samselikoff any update on this thing, ready to merge I believe? Anything else I can help on to make Mirage folder sharable between apps in an addon? Thanks

File: addon/orm/model.js [hunk 4]
```diff
@@ -79,6 +79,7 @@ class Model {
     }
 
     Object.keys(attrs).forEach(function(attr) {
+      this._definePlainAttribute(attr);
```
Comment (samselikoff) [inline addon/orm/model.js]: these shouldn't be part of this PR, could you rebase off master?

Comment (Leooo) [inline addon/orm/model.js]: rebased

**Comment by samselikoff** [comment]
Thanks for all your work on this!

**Comment by Leooo** [comment]
You're welcome @samselikoff, what are the next steps to be able to share the Mirage folder in an addon? Would be happy to work on it if you explain.

L

**Comment by samselikoff** [comment]
Tracking in #899
